### PR TITLE
subquery rfc

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

Closes #.

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: docs/rfcs/subquery_design.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+Authors: xudong.w
+
+RFC PR: #1373
+
+Datafusion Issue: #1209
+
+---
+
+### BackGround
```
Comment (Dandandan) [inline docs/rfcs/subquery_design.md:32]:
```suggestion
### Background
```

File: docs/rfcs/subquery_design.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+### Non-Goals
+1. Correlation subquery doesn't use a more complex solution to do decorrelation, for this design, I'll use nest-loop to do decorrelation.
+I'll provide some materials about more advanced decorrelation, and we can research them and optimize in subsequent iterations.
+2. Some unusual subqueries won't be supported, such as list subqueries.
+---
+
+### Survey
+
+#### postgresql
```
Comment (Dandandan) [inline docs/rfcs/subquery_design.md:56]:
```suggestion
#### PostgreSQL
```

File: docs/rfcs/subquery_design.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+- [NOT IN] -> [Anti] Semi Join
+- [ANY/SOME] -> Semi Join
+- [NOT EXISTS] -> [Anti] Semi Join
+
+**Note that not all sublinks can be pulled up**. Only some simple sublink queries can be pulled up. If contains aggregate functions, sublink won't be pulled up, then there will be a `SubPlan` in `Query Plan` and the upper execution plan and the subquery plan do nested loops to get the final result.
+
+Subqueries and sublinks have similar processing logic, except they are in different locations of queries
+
+
+#### cockroach
```
Comment (Dandandan) [inline docs/rfcs/subquery_design.md:68]:
```suggestion
#### CockroachDB
```

File: docs/rfcs/subquery_design.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+The cockroach's subquery design draws on the ideas in this [paper](https://dl.acm.org/doi/10.1145/1247480.1247598), you can see it to learn more details.
+That's the [decorrelation rules](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/norm/rules/decorrelate.opt) of the cockroach.
+
+#### materialize
```
Comment (Dandandan) [inline docs/rfcs/subquery_design.md:72]:
```suggestion
#### Materialize
```

**Comment by xudong963** [comment]
Thanks, @Dandandan , I am speeding up the completion of this design document

File: docs/rfcs/subquery_design.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+Materialize is a streaming system with incremental backend. This backend does not support nested loops, so it has to be able to handle all decorrelation cases. 
+Materialize's subquery implementation should be the most complete. This [paper](https://cs.emis.de/LNI/Proceedings/Proceedings241/383.pdf) is relevant to their implementation. The paper is a little indigestible, so I didn't deep into it.
+
+---
+
+### Design
+Let's turn our attention to Datafusion.
+
+There are two overall steps:
+1. transfer sqlparser subquery exprs to datafusion exprs.
```
Comment (hntd187) [inline docs/rfcs/subquery_design.md:82]: Do you mean something like an `Into<Expr>` impl or something of the sort on the datafusion side?

Comment (xudong963) [reply docs/rfcs/subquery_design.md:82]: `sql_expr_to_logical_expr` function in datafusion/src/sql/planner.rs

**Comment by alamb** [comment]
Marking PRs without activity in the last month as stale. I'll plan to close it in another month or so without activity, though feel free to reopen it when you have time to work on it)

**Comment by alamb** [comment]
Closing stale PRs. Please reopen (or open a new one) if you plan to keep working on this feature.

**Pull request closed without merging**